sound/soc/sof/ipc4-loader.c

```diff
@@ -60,67 +60,110 @@ static size_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev)
 		return -EINVAL;
 	}
 
+	/* return the extended manifest header length if no_parse is true */
+	if (no_parse)
+		return ext_man_hdr->len;
+
 	dev_info(sdev->dev, "Loaded firmware version: %u.%u.%u.%u\n",
 		 fw_header->major_version, fw_header->minor_version,
 		 fw_header->hotfix_version, fw_header->build_version);
 	dev_dbg(sdev->dev, "Firmware name: %s, header length: %u, module count: %u\n",
 		fw_header->name, fw_header->len, fw_header->num_module_entries);
 
-	ipc4_data->fw_modules = devm_kmalloc_array(sdev->dev,
-						   fw_header->num_module_entries,
-						   sizeof(*fw_module), GFP_KERNEL);
-	if (!ipc4_data->fw_modules)
-		return -ENOMEM;
+	/*
+	 * allocate memory for the number of modules in the base FW + the max libraries
+	 * supported by the firmware
+	 */
+	if (!lib_index) {
+		/* max_fw_libs includes the base firmware */
+		int num_total_modules = fw_header->num_module_entries + ipc4_data->max_fw_libs - 1;
+
+		ipc4_data->fw_modules = devm_kmalloc_array(sdev->dev, num_total_modules,
+							   sizeof(*fw_module), GFP_KERNEL);
+		if (!ipc4_data->fw_modules)
+			return -ENOMEM;
+
+		ipc4_data->module_uuids = devm_kcalloc(sdev->dev, num_total_modules,
+						       sizeof(guid_t), GFP_KERNEL);
+		if (!ipc4_data->module_uuids)
+			return -ENOMEM;
+	} else {
+		char *fw_name = kasprintf(GFP_KERNEL, "%s", fw_header->name);
+
+		if (!fw_name)
+			return -ENOMEM;
+
+		/* set the FW name for the 3rd party modules */
+		ipc4_data->module_lib_info[lib_index].name = fw_name;
+	}
 
-	ipc4_data->num_fw_modules = fw_header->num_module_entries;
-	fw_module = ipc4_data->fw_modules;
+	fw_module = ipc4_data->fw_modules + ipc4_data->num_fw_modules;
 
 	fm_entry = (struct sof_man4_module *)((u8 *)fw_header + fw_header->len);
 	remaining -= fw_header->len;
 
 	if (remaining < fw_header->num_module_entries * sizeof(*fm_entry)) {
 		dev_err(sdev->dev, "Invalid num_module_entries %u\n",
 			fw_header->num_module_entries);
-		return -EINVAL;
+		ret = -EINVAL;
+		goto err;
 	}
 
 	fm_config = (struct sof_man4_module_config *)
 				(fm_entry + fw_header->num_module_entries);
 	remaining -= (fw_header->num_module_entries * sizeof(*fm_entry));
 	for (i = 0; i < fw_header->num_module_entries; i++) {
-		memcpy(&fw_module->man4_module_entry, fm_entry, sizeof(*fm_entry));
+		/* ignore modules with null UUID */
+		if (guid_is_null(&fm_entry->uuid)) {
+			fm_entry++;
+			continue;
+		}
 
 		if (fm_entry->cfg_count) {
 			if (remaining < (fm_entry->cfg_offset + fm_entry->cfg_count) *
 			    sizeof(*fm_config)) {
 				dev_err(sdev->dev, "Invalid module cfg_offset %u\n",
 					fm_entry->cfg_offset);
-				return -EINVAL;
+				ret = -EINVAL;
+				goto err;
 			}
 
 			/* a module's config is always the same size */
 			fw_module->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
 
+			guid_copy(&ipc4_data->module_uuids[i + ipc4_data->num_fw_modules],
```

Comment:
`&ipc4_data->module_uuids[ipc4_data->num_fw_modules + i]` for visually more telling?
If the point is that we start updating the uuids after the num_fw_modules.